src/librbd/mirror/snapshot/RemoveImageStateRequest.cc

```diff
@@ -0,0 +1,127 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/mirror/snapshot/RemoveImageStateRequest.h"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/Utils.h"
+#include "librbd/mirror/snapshot/Types.h"
+#include "librbd/mirror/snapshot/Utils.h"
+
+#define dout_subsys ceph_subsys_rbd
+
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::mirror::snapshot::RemoveImageStateRequest: " \
+                           << this << " " << __func__ << ": "
+
+namespace librbd {
+namespace mirror {
+namespace snapshot {
+
+using librbd::util::create_rados_callback;
+
+template <typename I>
+void RemoveImageStateRequest<I>::send() {
+  get_object_count();
+}
+
+
+template <typename I>
+void RemoveImageStateRequest<I>::get_object_count() {
+  CephContext *cct = m_image_ctx->cct;
+
+  auto oid = util::image_state_object_name(m_image_ctx, m_snap_id, 0);
+  ldout(cct, 20) << oid << dendl;
+
+  librados::ObjectReadOperation op;
+  op.read(0, ImageStateHeader::length(), &m_bl, nullptr);
+
+  librados::AioCompletion *comp = create_rados_callback<
+    RemoveImageStateRequest<I>,
+    &RemoveImageStateRequest<I>::handle_get_object_count>(this);
+  int r = m_image_ctx->md_ctx.aio_operate(oid, comp, &op, nullptr);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void RemoveImageStateRequest<I>::handle_get_object_count(int r) {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << "r=" << r << dendl;
+
+  if (r < 0) {
+    lderr(cct) << "failed to read image state object: " << cpp_strerror(r)
+               << dendl;
+    finish(r);
+    return;
+  }
+
+  ImageStateHeader header(1);
+  try {
+    using ceph::decode;
+    decode(header, m_bl);
+  } catch (const buffer::error &err) {
+    lderr(cct) << "failed to decode image state object header" << dendl;
+    // still try to remove it
+  }
+  m_object_count = header.object_count;
+
+  remove_object();
+}
+
+template <typename I>
+void RemoveImageStateRequest<I>::remove_object() {
+  CephContext *cct = m_image_ctx->cct;
+
+  auto oid = util::image_state_object_name(m_image_ctx, m_snap_id,
+                                           m_object_index);
+  ldout(cct, 20) << oid << dendl;
+
+  librados::ObjectWriteOperation op;
+  op.remove();
+
+  librados::AioCompletion *comp = create_rados_callback<
+    RemoveImageStateRequest<I>,
+    &RemoveImageStateRequest<I>::handle_remove_object>(this);
+  int r = m_image_ctx->md_ctx.aio_operate(oid, comp, &op);
+  ceph_assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+void RemoveImageStateRequest<I>::handle_remove_object(int r) {
+  CephContext *cct = m_image_ctx->cct;
+  ldout(cct, 20) << "r=" << r << dendl;
+
+  if (r < 0 && r != -ENOENT) {
+    lderr(cct) << "failed to remove image state object: " << cpp_strerror(r)
+               << dendl;
+    finish(r);
+    return;
+  }
+
+  m_object_index++;
```

Comment:
Go in reverse order to ensure an interrupted removal can get restarted.

Reply:
I thought it could be restarted anyway as I made it to ignore ENOENT, but ok.

Reply:
Ah, never mind. Sure it couldn't be restarted as we could not reed the object header. Actually the reason why I wanted to remove from 0 was to make sure nobody starts reading a state that is in a process of removal. But this does not look like important, so I will change the order.